### ISSUE TITLE
[rllib]: regression bug. Catastrophic drop in SAC performance after updating from 1.2

### ISSUE THREAD

**Issue by MischaPanch** [ray-project/ray#17844]
### What is the problem?

This is a regression bug. After updating from ray 1.2.0 to later versions, SAC (with torch framework, torch version 1.8.1) stopped converging at all on my environment. See the learning curve below
 
I ruled out the following possibilities:

1) This issue is not related to the removing of state_preprocessor option in 1.3.
2) It is not due to transcendental dependencies coming from installing different versions of ray at different moments of time. I made sure that the only difference in my experiments is the version of ray, not any other package.
3) It is not caused by randomness. The learning curves and convergence are very stable, I get the same results each time I run the experiment.

Btw, due to the non-pinned version of aioredis a clean install of any earlier version of ray will lead to a "broken" python environment, in the sense that ray will not run there. So, in order to reproduce this issue, one first needs to adjust the environment by downgrading aioredis.

I suspect that the problem is due to changes in this PR: https://github.com/ray-project/ray/pull/13522 since it is the only major change related to SAC from 1.2 to 1.3. It might be something else, of course. I am happy to try to fix this or identify the cause if you could give me some guidance.

One strange thing: the bad convergence curves lead to agents that are worse (or close to) than random actions. Could it be as simple as just a sign error? I would have suspected that someone would have the same issue then but maybe I am wrong :)
 
![image](https://user-images.githubusercontent.com/35432522/129448242-b311fd46-4c13-41c2-bb7a-26228e97801c.png)

Orange curve is ray 1.2.0 The other curves are 1.3.0, 1.4.0 and 1.5.2

### Reproduction
Unfortunately, I cannot share the code leading to this since it is proprietary. I have not tried on another environment but from the behavior I observe I am very certain that this is a bug in ray (the only difference in experiments being the version).

**Comment by sven1977** [comment]
Hey @MischaPanch , could you check, whether your model is actually the same for those comparisons. For example, just print out the number of trainable weights in both cases (before the drop and after). Would like to understand this issue better.

**Comment by sven1977** [comment]
Also, what's your observation space?

**Comment by MischaPanch** [comment]
Hey @sven1977, thanks for the answer! I am constructing the model with ray, using the same config for all experiments. So, if the model is not the same that would also be a bug in rllib ^^.  Here the agent config (excluding the env config) passed to tune

```python
agent_config = {
    ...
    "framework": "torch",
    "observation_filter": "MeanStdFilter",
    "Q_model": {
        "fcnet_activation": "relu",
        "fcnet_hiddens": [128, 256, 256],
    },
    "policy_model": {
        "fcnet_activation": "relu",
        "fcnet_hiddens": [128, 256, 256],
    },
    "timesteps_per_iteration": 500, 
}
```

Essentially, I just run
```
trainer = sac.SACTrainer
tune.run(trainer, config=agent_config, ...)
```

My observation space is just a flat vector. One slightly special thing is that because I have a partially observed process, I add previous observations and actions to the observation passed to the RL agent. For that I implemented the following wrapper which I use together with gym's FrameStack

```python
class AddPreviousActionWrapper(Wrapper):
    """
    This wrapper is suitable for environments which correspond to partially observed markov decision processes. Then
    adding the previous action to the observation might be very beneficial for the training process.

    Note: If used in conjunction with gym's FrameStack wrapper (i.e. if FrameStack is applied after applying this wrapper),
    the combination becomes a TrajectoryViewWrapper, where the state consists of the last N observations and actions.
    """

    def __init__(self, env):
        super().__init__(env)
        self.previous_action = None
        self.observation_space = concatenate_boxes(
            [self.observation_space, self.action_space]
        )

    def get_enhanced_observation(self, observation: np.ndarray):
        return np.concatenate([observation, self.previous_action])

    def reset(self, **kwargs):
        observation = self.env.reset(**kwargs)
        self.previous_action = self.action_space.sample()
        return self.get_enhanced_observation(observation)

    def step(self, action: np.ndarray):
        observation, reward, done, info = self.env.step(action)
        observation = self.get_enhanced_observation(observation)
        self.previous_action = action
        return observation, reward, done, info
```

**Comment by MischaPanch** [comment]
I will try training the same model with tensorflow in different versions of ray and will write the results here soon (probably later today)

**Comment by MischaPanch** [comment]
Update: training with tensorflow. Somehow surprisingly, training with tensorflow also behaves worse than pytorch in version 1.2.0. In 1.3.0 it gets even worse

![image](https://user-images.githubusercontent.com/35432522/130781473-f2b40e73-02ee-452a-a150-d898c7666358.png)
Orange curve is torch on 1.2.0, pink is tf2 on 1.2.0 and grey is tf2 on 1.3.0.

I used the newest versions of tensorflow and tensorflow-probability

**Comment by sven1977** [comment]
Hey @MischaPanch , thanks for posting the results. My current suspicion is that the models being constructed are different between the versions.
Can you do the following for the working and the non-working releases?

```
# New version (broken one)
trainer = SACTrainer(config=[your config], env=[your env])
pol = trainer.get_policy()
model = pol.model

model.q_net.base_model.summary()
model.action_model.base_model.summary()

# Old version (working one):
trainer = SACTrainer(config=[your config], env=[your env])
pol = trainer.get_policy()
model = pol.model

model.q_net.summary()
model.action_model.summary()
```

**Comment by sven1977** [comment]
Some more questions:
- Your `concatenate_boxes` method: Does it return a gym.spaces.Tuple space?
- What are the shapes of these two spaces: self.observation_space, self.action_space?

**Comment by sven1977** [comment]
I found an issue with n-step > 1, affecting possibly SAC, DDPG, DQN and their derivatives (e.g. APEX).
Here is a PR that fixes the problem. We need to confirm that this is the only reason for the drop in performance.
https://github.com/ray-project/ray/pull/18358

**Comment by sven1977** [comment]
@MischaPanch ^

**Comment by MischaPanch** [comment]
Hi @sven1977 , sorry for the delayed response.

1. `concatenate_boxes` returns a box, see the implementation below (for the sake of completeness). I only allow concatenation of flat vectors, so my observation is always flat. The actions for me are either 1-dim., 2-dim. or 3-dim. vectors. In the experiment shown above they were one dimensional.
2. The models don't have the `.summary()` method nor the `base_model` attribute (I am using pytorch) but their repr seems good enough. I post the outputs below.
As you see, there is quite some difference! I haven't tried to dive into that yet so I don't know how the `ComplexInputNetwork` differs from the previous `Sequential` model but the layers do seem a bit off.  Hidden layers being an empty sequence seems suspicious, right? Also, the action now has a `_value_branch` layer. I have clicked through the ray 1.2.0 code before and I believe the torch modelV2 value network that was always created by default was never actually trained or used in the SAC implementation (I may be wrong there). Can't say how it is handled in >= 1.3.0 though
3. I had the suspicion that initalization of weights was to blame. For the action network the weights seem to be initialized in the same fashion in ray 1.2.0 and in 1.3.0. For the q-net, the parameters are entirely different because the whole net is different, so no clear comparison is possible. However, I noticed that the non-zero weights in the new q_net were 1-2 orders of magnitude smaller than in the old one. 
4. I am happy do help you with diagnosing this issue. Seeing how the q_net is so different, should I still try to run the training on the code in your PR?

Here the old models:
```
Sequential(
  (q_hidden_0): SlimFC(
    (_model): Sequential(
      (0): Linear(in_features=193, out_features=128, bias=True)
      (1): ReLU()
    )
  )
  (q_hidden_1): SlimFC(
    (_model): Sequential(
      (0): Linear(in_features=128, out_features=256, bias=True)
      (1): ReLU()
    )
  )
  (q_hidden_2): SlimFC(
    (_model): Sequential(
      (0): Linear(in_features=256, out_features=256, bias=True)
      (1): ReLU()
    )
  )
  (q_out): SlimFC(
    (_model): Sequential(
      (0): Linear(in_features=256, out_features=1, bias=True)
    )
  )
)


Sequential(
  (action_0): SlimFC(
    (_model): Sequential(
      (0): Linear(in_features=192, out_features=128, bias=True)
      (1): ReLU()
    )
  )
  (action_1): SlimFC(
    (_model): Sequential(
      (0): Linear(in_features=128, out_features=256, bias=True)
      (1): ReLU()
    )
  )
  (action_2): SlimFC(
    (_model): Sequential(
      (0): Linear(in_features=256, out_features=256, bias=True)
      (1): ReLU()
    )
  )
  (action_out): SlimFC(
    (_model): Sequential(
      (0): Linear(in_features=256, out_features=2, bias=True)
    )
  )
)
```

And here the new ones (v 1.3.0)
```
ComplexInputNetwork(
  (post_fc_stack): FullyConnectedNetwork(
    (_hidden_layers): Sequential()
    (_value_branch_separate): Sequential()
    (_value_branch): SlimFC(
      (_model): Sequential(
        (0): Linear(in_features=193, out_features=1, bias=True)
      )
    )
  )
  (logits_layer): SlimFC(
    (_model): Sequential(
      (0): Linear(in_features=193, out_features=1, bias=True)
    )
  )
  (value_layer): SlimFC(
    (_model): Sequential(
      (0): Linear(in_features=193, out_features=1, bias=True)
    )
  )
)

FullyConnectedNetwork(
  (_logits): SlimFC(
    (_model): Sequential(
      (0): Linear(in_features=256, out_features=2, bias=True)
    )
  )
  (_hidden_layers): Sequential(
    (0): SlimFC(
      (_model): Sequential(
        (0): Linear(in_features=192, out_features=128, bias=True)
        (1): ReLU()
      )
    )
    (1): SlimFC(
      (_model): Sequential(
        (0): Linear(in_features=128, out_features=256, bias=True)
        (1): ReLU()
      )
    )
    (2): SlimFC(
      (_model): Sequential(
        (0): Linear(in_features=256, out_features=256, bias=True)
        (1): ReLU()
      )
    )
  )
  (_value_branch): SlimFC(
    (_model): Sequential(
      (0): Linear(in_features=256, out_features=1, bias=True)
    )
  )
)
```

Here my implementation of `concatenate_boxes`.

```python
def concatenate_boxes(boxes: List[Box]):
    result_lows = []
    result_highs = []
    for b in boxes:
        if len(b.shape) != 1:
            raise Exception(f"Can only concatenate flat boxes but got shape {b.shape}")
        result_lows.extend(b.low)
        result_highs.extend(b.high)
    return Box(np.array(result_lows), np.array(result_highs))
```
